node/external/transactionAPI/gasUsedAndFeeProcessor.go

```diff
@@ -57,10 +57,14 @@ func (gfp *gasUsedAndFeeProcessor) computeAndAttachGasUsedAndFee(tx *transaction
 		tx.GasUsed = big.NewInt(0).Div(initialTotalFee, big.NewInt(0).SetUint64(tx.GasPrice)).Uint64()
 	}
 
+	isHashForInnerTxActive := gfp.enableEpochsHandler.IsFlagEnabledInEpoch(common.HashForInnerTransactionFlag, tx.Epoch)
 	hasRefundForSender := false
 	totalRefunds := big.NewInt(0)
 	for _, scr := range tx.SmartContractResults {
-		if !scr.IsRefund || scr.RcvAddr != tx.Sender {
+		isScrRefundForRelayer := scr.IsRefund && (scr.RcvAddr == tx.RelayerAddress)
+		shouldConsiderRefundSentToRelayer := isScrRefundForRelayer && isHashForInnerTxActive
+		isScrForTxSender := scr.RcvAddr == tx.Sender
+		if !scr.IsRefund || (!isScrForTxSender && !shouldConsiderRefundSentToRelayer) {
```

Comment:
this is also very hard to understand what happens before and after the activation. Could it be refactored so that the before activation remains the same like before, and after the activation it has another logic?

Reply:
Run Tests:
mx-chain-simulator-go: fix-failed-test
mx-chain-testing-suite: link_inner_tx_to_results